### PR TITLE
feat: implement GetUser API and activated status

### PR DESCRIPTION
**Pull request by howar31**
# Issue
- New API: GetUser
- New migration: activated field in users table
- Modify API: add activated in login response
- Modify API: set activated after set interests and donate step 2

# Notice

# Dependency

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1204378944314708
  - https://app.asana.com/0/0/1204692921007278

### CODE REVIEW

File: doc/user.apib [hunk 1]
```diff
@@ -1,13 +1,14 @@
...
-    + name: Full Name (string) - The full name of the user
-    + activated: true (boolean) - The activated status of the user (e.g. true / false)
-    + read_preference (array) - Subscribed topics
-    + maillist (array) - Subscribed mail list
++ user_id: 123 (string) - The unique identifier of the user
++ email: example@email.com (string) - The email address of the user
++ name: Full Name (string) - The full name of the user
++ roles: explorer, action_taker, trailblazer (array) - The roles of the user
++ registration_date: `2018-01-24T22:16:48Z` (string) - The registration date of the user
++ activated: `2023-06-01T01:23:45Z` (string) - The activated status of the user. If the user is not activated, the value is null.
```
Comment (Aylie-Chou) [inline doc/user.apib:9]: According to [design doc](https://www.notion.so/twreporter/User-API-6fb09be7357745dbbabf68d0fbc278ad?pvs=4), the type of `activated` column would be `boolean`.
Do we have any specific reason to change its type ?

File: storage/user.go [hunk 2]
```diff
@@ -17,8 +17,9 @@ import (
 func (gs *GormStorage) GetUserByID(userID string) (models.User, error) {
 	user := models.User{}
 
-	// SELECT * FROM users WHERE ID = $userID
-	if err := gs.db.First(&user, "id = ?", userID).Error; err != nil {
+	// SELECT * FROM users WHERE ID = $userID and join roles table by users_roles table
+	err := gs.db.Preload("Roles").First(&user, "id = ?", userID).Error
```
Comment (Aylie-Chou) [inline storage/user.go:21]: According to api doc, the return data of `user.role` would be role name string.
Looks like we could preload `Roles.name` only here ?

File: storage/user.go [hunk 3]
```diff
@@ -17,8 +17,9 @@ import (
 func (gs *GormStorage) GetUserByID(userID string) (models.User, error) {
```
Comment (Aylie-Chou) [inline storage/user.go:17]: As previous discussion, the return data `user.maillist` would need to be mapped to maillist keys.

File: doc/user.apib [hunk 4]
```diff
@@ -1,13 +1,14 @@
...
-+ data (object, required) - The user object
-    + user_id: 123 (string) - The unique identifier of the user
-    + email: example@email.com (string) - The email address of the user
-    + name: Full Name (string) - The full name of the user
-    + activated: true (boolean) - The activated status of the user (e.g. true / false)
-    + read_preference (array) - Subscribed topics
-    + maillist (array) - Subscribed mail list
++ user_id: 123 (string) - The unique identifier of the user
++ email: example@email.com (string) - The email address of the user
++ name: Full Name (string) - The full name of the user
```
Comment (Aylie-Chou) [inline doc/user.apib:6]: As previous discussion, please help to change `name` into firstname, lastname.

File: controllers/donation.go [hunk 5]
```diff
@@ -629,18 +629,28 @@ func (mc *MembershipController) CreateAPeriodicDonationOfAUser(c *gin.Context) (
...
+			ID:          tokenDonation.ID,
 			OrderNumber: tokenDonation.OrderNumber,
-			Type: globals.TokenDonationType,
+			Type:        globals.TokenDonationType,
 		},
 	}
 	go publishToNeticrm(ms)
 
+	// Update the user's activated time to now
+	matchedUser, err := mc.Storage.GetUserByEmail(reqBody.Cardholder.Email)
```
Comment (Aylie-Chou) [inline controllers/donation.go:645]: Could we update activated value with concurrency in order not to block `CreateAPeriodicDonationOfAUser` ?

**Review by Aylie-Chou — CHANGES_REQUESTED** [review]
Mostly looking good.
I left some comments about previous discussion or mismatch deisign, please have a look.

File: doc/user.apib (hunk 1 above)
Comment (howar31) [reply doc/user.apib:9]: As the discussion, it was decided to change the boolean data type to timestamp in order to store more information without significantly impacting performance.

File: storage/user.go (hunk 2 above)
Comment (howar31) [reply storage/user.go:21]: We loaded `name` and `name_en` here.  I think it's ok to preload Roles.

File: storage/user.go (hunk 3 above)
Comment (howar31) [reply storage/user.go:17]: according to Asana tasks, we return email, registered date and roles.  maillist is not implemented in this commit.

File: doc/user.apib (hunk 4 above)
Comment (howar31) [reply doc/user.apib:6]: done

File: controllers/donation.go (hunk 5 above)
Comment (howar31) [reply controllers/donation.go:645]: done

Comment (Aylie-Chou) [reply controllers/donation.go:645]: if the return data of `GetUserByEmail` is only for updating user's activated status, could we also move it into the concurrent function ?

File: doc/user.apib [hunk 6]
```diff
@@ -1,13 +1,15 @@
...
-    + email: example@email.com (string) - The email address of the user
-    + name: Full Name (string) - The full name of the user
-    + activated: true (boolean) - The activated status of the user (e.g. true / false)
-    + read_preference (array) - Subscribed topics
-    + maillist (array) - Subscribed mail list
++ user_id: 123 (string) - The unique identifier of the user
++ email: example@email.com (string) - The email address of the user
++ fisrt_name: First Name (string) - The first name of the user
++ last_name: Last Name (string) - The last name of the user
++ roles: explorer, action_taker, trailblazer (array) - The roles of the user
```
Comment (Aylie-Chou) [inline doc/user.apib:8]: looks like the return value of `roles` is array of object. Could we specify what info would be contianed in the object here ?

File: storage/user.go (hunk 3 above)
Comment (Aylie-Chou) [reply storage/user.go:17]: updated in commit [3285298](https://github.com/twreporter/go-api/pull/552/commits/3285298385862ca61aa94dd4e982b6e571193df2).

**Review by Aylie-Chou — CHANGES_REQUESTED** [review]
Thanks for the quick resolving for the `getUser` return data.
Please have a look at the comments & feel free for any discussion.

File: controllers/donation.go (hunk 5 above)
Comment (howar31) [reply controllers/donation.go]: done

File: doc/user.apib (hunk 6 above)
Comment (howar31) [reply doc/user.apib]: done

**Review by Aylie-Chou — APPROVED** [review]
LGTM 👍

**Pull request merged**